Implement Roadmap and Upgrade layouts

wackerow opened the pull request:
## Description
- Implements Roadmap layout and related components
- To complete the `/roadmap` section, the Upgrade layout was also implemented along with its related components

## Paths enabled
Roadmap layout:
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/future-proofing
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/scaling
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/user-experience
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/security

Update layout:
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/beacon-chain
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/merge

Static layout:
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/account-abstraction
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/danksharding
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/pbs
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/secret-leader-election
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/single-slot-finality
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/statelessness
- https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap/verkle-trees

wackerow committed:
initialize Roadmap pages

wackerow committed:
Merge branch 'use-cases' into roadmap

wackerow committed:
clean up

wackerow committed:
Implement Upgrades layout and related components

wackerow committed:
Merge branch 'use-cases' into roadmap-layout

wackerow committed:
Merge branch 'use-cases' into roadmap-layout

wackerow committed:
Merge branch 'use-cases' into roadmap-layout

wackerow committed:
Merge branch 'use-cases' into roadmap-layout

wackerow committed:
enable remaining /roadmap paths

wackerow committed:
add /secret-leader-election, sort paths

wackerow committed:
Merge branch 'dev' into roadmap-layout

wackerow commented on public/upgrades/merge-infographic-bg.svg:
Relocated to / rebuilt as `@/components/MergeInfographic/Background.tsx`

pettinarip commented on public/content/roadmap/index.md:
```diff
@@ -66,7 +66,7 @@ The roadmap is mostly the result of years of work by researchers and developers
 </RoadmapImageContent>
 
 <InfoBanner mb={8}>
-  <h4 style="margin-top: 0">What was ETH2?</h4>
```
I would suggest keeping these changes out of the scope of this PR since we are doing updates from the Gatsby repo to this repo. So, this will create more conflicts.

I'd recommend any "out of scope" refactor or fix to be done in the Gatsby repo first. Then we can bring those changes in the regular syncs without the need to track any conflict.

wackerow replied:
@pettinarip This was a result of compiler errors I was getting. The build failed because it was expecting an object for the `style` attribute, not a string.

wackerow replied:
Otherwise I agree =) If you know which ts setting we can tweak to fix this, we can use that and revert this.

I could post a PR to the Gatsby repo, but it's not required there, so given it's only needed here and seems to be breaking without it I went with just refactoring them.

@minimalsm On this note... can you recall if this would cause any Crowdin issues? I don't think it should, but can make the syntax a bit more complex looking.

pettinarip replied:
Ah okok, thats ok then. If we are fixing an issue that only happens in this repo then its ok. Its part of the migration.

pettinarip commented on src/components/MergeInfographic/Background.tsx:
```diff
@@ -0,0 +1,91 @@
+    fill: "currentColor",
+  },
+  path: [
+    <g clip-path="url(#clip0_10_161)">
```
All these attrs should be camelCased
```suggestion
    <g clipPath="url(#clip0_10_161)">
```

wackerow replied:
Hah, that last step I sometimes forget when converting .svg's to React =)... patched.

pettinarip commented on src/components/MergeInfographic/Background.tsx:
Is this refactor done for any problem specific to Next? or is more a refactor to make it cleaner?

wackerow replied:
Hm 🤔  I was having trouble importing this directly as the `.svg` file... I was already quite familiar with how to use `createIcon` so decided to just pull this into the `/components` folder. It's an incomplete SVG anyway (only the background) so figured we're not losing much by just ditching the .svg file in lieu of this.

pettinarip replied:
Sounds good 👍🏼 Im good with this change.

However, I'm wondering what type of issue you had when importing an `.svg`. You can not use it as a component, right?

wackerow replied:
Yeah... I was having trouble out-of-box being able to just import it and use it as a component... I know I've implemented this on other repos before in various ways, but given we already have been using `createIcon` I figured that would work. 

But just for the info:

```ts
import Background from "@/../public/assets/merge-infographic-bg.svg"
// ...
        <Background
          aria-hidden="true"
          position="absolute"
          width="100%"
          height="100%"
          zIndex={0}
        />

// Resulted in:
Type '{ "aria-hidden": "true"; position: string; width: string; height: string; zIndex: number; }' is not assignable to type 'IntrinsicAttributes & SVGProps<SVGElement>'.
  Property 'position' does not exist on type 'IntrinsicAttributes & SVGProps<SVGElement>'.ts(2322)
// Amongst other typing errors for these props
```


```ts
import Background from "@/../public/assets/merge-infographic-bg.svg"
// ...
        <Icon
          as={Background}
          aria-hidden="true"
          position="absolute"
          width="100%"
          height="100%"
          zIndex={0}
        />

// Resulted in:
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.
```

pettinarip replied:
Thanks for the context.

The problem with the first one, I believe is because you are trying to use chakra props on it while it is just an svg element.

You should "chakrify" it first as [we are doing in the other repo](https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/MergeInfographic.tsx#L10).

```ts
const ChakraBackground = chakra(Background)
```

wackerow replied:
🤦 Brilliant hah... yeah that makes sense. Personally think we can still just keep this as-is, but thanks for mentioning that, will keep in mind in the future

pettinarip commented on src/components/Roadmap/RoadmapActionCard.tsx:
```diff
@@ -31,53 +27,13 @@ const RoadmapActionCard: React.FC<IProps> = ({
+    futureProofing: "/roadmap/roadmap-future.png",
+    scaling: "/roadmap/roadmap-transactions.png",
+    security: "/roadmap/roadmap-security.png",
+    userExperience: "/roadmap/roadmap-ux.png",
```
These images should be imported as normal imports.

Check [this section](https://nextjs.org/docs/app/building-your-application/optimizing/images#local-images) in the next docs. Basically, by importing it statically Next can calculate the width and height for you.

Right now, if Im not wrong, these images are not being displayed https://deploy-preview-38--ethereum-org-fork.netlify.app/roadmap

wackerow replied:
Yeah, I pushed an update around the time you were reviewing this that added back some explicit height/width numbers, based on the graphql queries we were using on Gatsby, which does fix it, but is manual. 

I just switched to the approach you showed and it removed the need to hard-code these ❤️ 👍 

---

@pettinarip @nhsz Side note... things in `public` are easy to reference in the DOM, but slightly messier to import. We can simplify the imports using this syntax for example:

```ts
import userExperience from "@/../public/roadmap/roadmap-ux.png"
```

If we add the following to `compilerOptions.paths` in `tsconfig.json` we could use `@/public/` to simplify further any imports from the public folder:

```ts
`"@/public/*": ["./public/*"]`
```

^ Let me know if you have thoughts on this... easy patch that could be applied separately if we want it. Going to use `@/../public/` format for now to avoid using relative paths.

pettinarip replied:
I'm down with adding that alias.

However, shouldn't the new import syntax be
- `"@/public/roadmap/roadmap-ux.png"` instead of
- `"@/../public/roadmap/roadmap-ux.png"`?

wackerow replied:
Yepp!.... but only IF we add that line to the `tsconfig` =)... If we're good with that pattern, I'll update `tsconfig` and switch `@/../public/` to `@/public/`

wackerow replied:
Just pushed a commit to update this using `@/public`

pettinarip commented on src/components/Roadmap/RoadmapImageContent.tsx:
```diff
@@ -1,46 +1,24 @@
+      <Stack mb={4}>{children}</Stack>
+      <Center>
+        <Image
+          src="wallet.png"
```
import the image
```tsx
import walletImage from '/wallet.png'

<Image src={walletImage} ...>
```

pettinarip commented on src/components/SectionNav.tsx:
```diff
@@ -23,16 +23,18 @@ const SectionNav: React.FC<IProps> = ({ children }) => (
     <Flex align="flex-start">
       <Emoji text=":point_right:" mb={4} />
-      <OldHeading
+      <Heading
```
For logistics reasons, lets not include these types of changes in this migration. If you'd like to do them, lets do them first in the Gatsby repo.

wackerow replied:
Oops, intended to revert... however, this component in particular is peculiar as it doesn't appear to be used _anywhere_ in the repo, but it _is_ imported and exported by every template/layout. No sign of it being used in a single markdown file. In another PR, (#45 I believe) I simply removed this component.

@pettinarip @nhsz Am I missing something or can we ignore this component and allow it to be removed in the other PR?

pettinarip replied:
Oh yes haha, seems that we are not using it. I've also double-checked this in the Gatsby repo and we are not using it. I'd suggest to just remove it in this PR as I need to do a sync PR soon (today).

wackerow replied:
Okay can do!

pettinarip commented on src/lib/interfaces.ts:
```diff
@@ -62,3 +62,19 @@ export interface PageContent {
   tocItems: Array<ToCItem>
 }
+
+export interface RequiredFrontmatter {
```
👍🏼 nice

wackerow replied:
See #42 for more advanced iteration on this concept.

pettinarip commented on src/lib/types.ts:
```diff
@@ -12,6 +12,7 @@ export type AppPropsWithLayout = AppProps & {
 }
 
+// TODO: Build out Frontmatter schemas
 export type Frontmatter = { [key: string]: any }
```
So, the `RequiredFrontmatter` is only being used by the new layouts created in this PR, no? the Static is still using this other type.

wackerow replied:
Correct... I started playing with the front matter schema in this PR, but kept it to the new layouts... See #42 for the schema I'm proposing which is rolled out more widely in that PR.

pettinarip commented on src/layouts/Upgrade.tsx:
```diff
@@ -1,3 +1,433 @@
+  // TODO some `gitLogLatestDate` are `null` - why?
+  // const lastUpdatedDate = parent.fields
+  //   ? parent.fields.gitLogLatestDate
+  //   : parent.mtime
```
Are you planning on doing this in a separate PR? could we create a new card to keep track of it?

wackerow replied:
@pettinarip Good call... I thought we needed more work for this but the `lastUpdatedDate` has already been implemented... just pushed some commits that re-implement this.

I have a separate task (card in Notion) tee'd up for other GitHub fetches we'll need to do beyond the last updated date.

pettinarip replied:
Awesome. Marking this as done for now.

wackerow committed:
fix images for roadmap components

pettinarip reviewed (commented):
Nice job @wackerow 

Left a few comments.

Would like to know from your side what TODOs or remaining work is left from this PR that you think we should keep track of or create a task, in order to consider these layouts fully migrated.

wackerow committed:
fix camelCase svg attributes

wackerow committed:
import roadmap image assets from public

wackerow committed:
use lastUpdatedDate in Upgrade

updated interface using optional lastUpdatedDate, but this should be hardened in the Front matter schema PR #42

wackerow committed:
enable BeaconChainActions and related

wackerow committed:
rm unused SectionNav component

wackerow committed:
add @/public ts shortcut

pettinarip commented on src/components/CardList.tsx:
```diff
@@ -1,52 +1,44 @@
+import { BaseLink } from "@/components/Link"
+import { Image } from "@/components/Image"
+import * as url from "@/lib/utils/url"
+import type { StaticImageData } from "next/image"
```
nitpick: import order

wackerow replied:
nitpick of mine as well, good catch =)

pettinarip commented on src/components/CardList.tsx:
```diff
@@ -1,52 +1,44 @@
   link?: string
   id?: string
-  image?: IGatsbyImageData
+  image?: string | StaticImageData
```
This is good but I think there is a better way of declaring this type:
```ts
import { Image, type ImageProps } from "@/components/Image"

export type CardListItem = {
  ...
  image?: ImageProps['src']
}
```
by reusing the same declared types of the custom component that we have defined. If, for some reason, we need to refactor our `Image` component types, we won't need to worry about these types definitions.

wackerow replied:
Okay yeah, I like that, can patch

pettinarip commented on src/components/Roadmap/RoadmapActionCard.tsx:
```diff
@@ -87,8 +47,8 @@ const RoadmapActionCard: React.FC<IProps> = ({
+        <Image
+          src={imgSrc}
           alt={alt}
           objectFit="contain"
```
`objectFit` prop is deprecated. We need to use the `style` prop from now on to specify that
```suggestion
          style={{
            objectFit: "contain",
          }}
```

pettinarip commented on src/components/Roadmap/RoadmapImageContent.tsx:
```diff
@@ -1,46 +1,27 @@
+          src={wallet}
+          alt="Ethereum Wallet"
+          width={350}
+          objectFit="contain"
```
use `style` prop

pettinarip commented on src/layouts/Roadmap.tsx:
```diff
@@ -1,3 +1,260 @@
+  ListItem,
+  Show,
+  SimpleGrid,
+  Text,
```
Noting that we are migrating the `Text` component from the `OldText`. Would recommend to keep using the previous one for this migration to avoid any styling regression.

Right now, there is some space missing between the Text and the buttons
![image](https://github.com/ethereum/ethereum-org-fork/assets/468158/b1b9708d-5117-49e4-a7b6-9542a0f371dd)

Consider doing those changes in the Gatsby repo to keep both repos in sync.

wackerow replied:
My bad again sorry, this may have still been prior to us discussing this... will revert to minimize changes.

pettinarip commented on src/layouts/Roadmap.tsx:
```diff
@@ -1,3 +1,260 @@
+          <Image
+            src={frontmatter.image}
+            alt={frontmatter.alt || ""}
+            objectFit="contain"
```
use `style` prop

pettinarip commented on src/layouts/Upgrade.tsx:
```diff
@@ -1,3 +1,417 @@
+          <Image
+            flex="1 1 100%"
+            maxW="min(100%, 816px)"
+            sx={{ objectFit: "cover" }}
```
replace this by using [the `style` prop](https://nextjs.org/docs/pages/api-reference/components/image#style)

pettinarip commented on src/lib/interfaces.ts:
```diff
@@ -61,4 +61,21 @@ export interface PageContent {
+}
+
+export interface UpgradeFrontmatter extends RequiredFrontmatter {
+  description?: string
```
nitpicking: looking into the Upgrades layout, seems like the `description` should be required
```suggestion
  description: string
```

wackerow replied:
Yes, it is... I'll patch here, but note this is also patched in the Front matter PR that focuses on all of this in more detail.

In the end, `description` may be a blank string in some files, but the field is _required_ in **all** layouts, along with `title` and `lang`... but yeah, see #42 =)

pettinarip commented on src/components/BeaconChainActions.tsx:
```diff
@@ -25,71 +22,61 @@ const H3: React.FC<{ children: React.ReactNode }> = ({ children }) => (
+// type BeaconQueryTypes = {
+//   beaconscan: ImageDataLike | null
+//   beaconchain: ImageDataLike | null
+// }
```
can we delete these commented lines?

wackerow committed:
refactor objectFit style prop

wackerow committed:
refactor image prop type

wackerow committed:
apply suggestions from code review

Co-Authored-By: Pablo Pettinari <pettinarip@gmail.com>

wackerow committed:
add BeaconChainActions back

wackerow committed:
Merge branch 'dev' into roadmap-layout

wackerow commented:
(Build breaking due to known GitHub API rate limiting)

wackerow committed:
Merge branch 'roadmap-layout' of https://github.com/ethereum/ethereum-org-fork into roadmap-layout

wackerow committed:
Merge branch 'dev' into roadmap-layout

wackerow committed:
fix style prop type

nhsz committed:
Merge branch 'dev' into roadmap-layout

nhsz commented on src/components/CardList.tsx:
```diff
@@ -1,52 +1,44 @@
   LinkBox,
   LinkOverlay,
-  StackProps,
+  type BoxProps,
```
```suggestion
BoxProps,
```

nhsz commented on src/components/CardList.tsx:
```diff
@@ -1,52 +1,44 @@
   LinkOverlay,
-  StackProps,
+  type BoxProps,
+  type StackProps,
```
```suggestion
StackProps,
```

nhsz commented on src/components/CardList.tsx:
```diff
@@ -58,7 +50,7 @@ const Card = (props: CardListItem & Omit<StackProps, "title" | "id">) => {
   return (
     <CardContainer {...rest}>
-      {image && <GatsbyImage image={image} alt={alt || ""} minW="20px" />}
+      {image && <Image src={image} alt={alt || ""} minW="20px" />}
```
```suggestion
      {image && <Image src={image} alt={alt ?? ""} minW="20px" />}
```

nhsz commented on src/layouts/Roadmap.tsx:
```diff
@@ -1,3 +1,260 @@
+          </TitleCard>
+          <Image
+            src={frontmatter.image}
+            alt={frontmatter.alt || ""}
```
```suggestion
            alt={frontmatter.alt ?? ""}
```

nhsz committed:
Update src/components/CardList.tsx

nhsz committed:
Update src/components/CardList.tsx

nhsz committed:
Update src/components/CardList.tsx

nhsz committed:
Update src/layouts/Roadmap.tsx
